From b5c008a330596a8461508cf1ae66a5cc3d64032e Mon Sep 17 00:00:00 2001
From: Karthik Kambatla <kasha@apache.org>
Date: Wed, 17 Dec 2014 12:03:16 -0800
Subject: [PATCH 1365/2848] YARN-2945. FSLeafQueue#assignContainer - document
 the reason for using both write and read locks.
 (Tsuyoshi Ozawa via kasha) (cherry picked from
 commit f2d150ea1205b77a75c347ace667b4cd060aaf40)

(cherry picked from commit 2d832ad2eb87e0ce7c50899c54d05f612666518a)

Change-Id: I084e127088bffd34b53c38c99825cce17224fb52
---
 .../scheduler/fair/FSLeafQueue.java                |    3 +++
 1 file changed, 3 insertions(+)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSLeafQueue.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSLeafQueue.java
index 22b1f7d..842d3c3 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSLeafQueue.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSLeafQueue.java
@@ -321,6 +321,9 @@ public Resource assignContainer(FSSchedulerNode node) {
     } finally {
       writeLock.unlock();
     }
+    // Release write lock here for better performance and avoiding deadlocks.
+    // runnableApps can be in unsorted state because of this section,
+    // but we can accept it in practice since the probability is low.
     readLock.lock();
     try {
       for (FSAppAttempt sched : runnableApps) {
-- 
1.7.9.5

